### PR TITLE
[RLlib] Test against failure of nodes, for example for practical use of spot instances

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

Spot instances pose a great way to save money on less time critical jobs.
This PR introduces a test that is expected to fail for now, testing whether PPO can pause and continue learning on a single node failure.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/tests/test_node_failure.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    def tearDown(self):
+        ray.shutdown()
+
+    def test_fail_on_node_failure(self):
+        # We tolerate failing workers and don't stop training
+        config = (
+            PPOConfig()
+            .rollouts(
+                num_rollout_workers=6,
+                ignore_worker_failures=False,
```
Comment (sven1977) [inline rllib/tests/test_node_failure.py:54]: Wait, how do these settings "tolerate failing workers" as indicated in the comment above?

Shouldn't it be `ignore_worker_failures=True` or so?

File: rllib/tests/test_node_failure.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+        # Fail with a node down, resource requirements not satisfied anymore
+        try:
+            ppo.step()
+        except RayActorError:
+            return
+        raise ValueError("Expected algorithm to not step here")
+
+    def test_continue_training_on_failure(self):
+        # We tolerate failing workers and don't stop training
```
Comment (sven1977) [inline rllib/tests/test_node_failure.py:83]: Ah, got it. Wrong comment above? :)

File: rllib/tests/test_node_failure.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+        # One step with a node down, resource requirements not satisfied anymore
+        ppo.step()
+
+        # Training should have proceeded without errors, but two workers missing
+        self.assertEqual(len(ppo.workers._worker_health_check()), 0)
+        self.assertEqual(len(ppo.workers._remote_workers), 4)
+
+    def test_recreate_workers_on_next_iter(self):
+        # We tolerate failing workers and don't stop training
```
Comment (sven1977) [inline rllib/tests/test_node_failure.py:117]: same

File: rllib/tests/test_node_failure.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        # One step with a node down, resource requirements not satisfied anymore
+        # step() should hang trying to restore the missing workers.
+        ppo.step()
+
+        # Workers should be back up, everything back to normal.
+        self.assertEqual(len(ppo.workers._worker_health_check()), 0)
+        self.assertEqual(len(ppo.workers._remote_workers), 6)
+
+    def test_wait_for_nodes_on_failure(self):
+        # We tolerate failing workers and don't stop training
```
Comment (sven1977) [inline rllib/tests/test_node_failure.py:162]: same

File: rllib/tests/test_node_failure.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
+# This workload tests RLlib's ability to recover from failing workers nodes
```
Comment (sven1977) [inline rllib/tests/test_node_failure.py:1]: Very nice!
Thanks for these new tests @ArturNiederfahrenhorst .

File: rllib/tests/test_node_failure.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+import time
+import unittest
+import threading
+from ray.exceptions import RayActorError
```
Comment (gjoliver) [inline rllib/tests/test_node_failure.py:5]: group the ray imports together, sort them alphabetically.

File: rllib/tests/test_node_failure.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+import ray
+from ray.cluster_utils import Cluster
+from ray._private.test_utils import get_other_nodes
+
+from ray.rllib.algorithms.ppo import PPO, PPOConfig
+
+num_redis_shards = 5
+redis_max_memory = 10 ** 8
+object_store_memory = 10 ** 8
+num_nodes = 3
```
Comment (gjoliver) [inline rllib/tests/test_node_failure.py:16]: constants should be ALL_CAP?

File: rllib/tests/test_node_failure.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+message = (
+    "Make sure there is enough memory on this machine to run this "
+    "workload. We divide the system memory by 2 to provide a buffer."
+)
+assert (
+    num_nodes * object_store_memory + num_redis_shards * redis_max_memory
+    < ray._private.utils.get_system_memory() / 2
+), message
```
Comment (gjoliver) [inline rllib/tests/test_node_failure.py:25]: just write the message like this?

```
assert (
    num_nodes * object_store_memory + num_redis_shards * redis_max_memory
    < ray._private.utils.get_system_memory() / 2
), (
    "Make sure there is enough memory on this machine to run this "
    "workload. We divide the system memory by 2 to provide a buffer."
)
```

File: rllib/tests/test_node_failure.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        node_to_kill = get_other_nodes(self.cluster, exclude_head=True)[0]
+        self.cluster.remove_node(node_to_kill)
+
+        # Check faulty worker indices
+        # All nodes have 2 CPUs and we require a driver thread + 5 rollout workers
+        self.assertEqual(len(ppo.workers._worker_health_check()), 2)
+        self.assertEqual(len(ppo.workers._remote_workers), 6)
+
+        # Fail with a node down, resource requirements not satisfied anymore
+        try:
```
Comment (gjoliver) [inline rllib/tests/test_node_failure.py:76]: just use self.assertRaise()?

File: rllib/tests/test_node_failure.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+            redis_port=None,
+            num_redis_shards=None,
+            num_cpus=2,
+            num_gpus=0,
+            object_store_memory=object_store_memory,
+            redis_max_memory=redis_max_memory,
+            dashboard_host="0.0.0.0",
+        )
+
+        # One step with a node down, resource requirements not satisfied anymore
```
Comment (gjoliver) [inline rllib/tests/test_node_failure.py:153]: resource requirements should still satisfy for this case, I think.

File: rllib/tests/test_node_failure.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            object_store_memory=object_store_memory,
+            redis_max_memory=redis_max_memory,
+            dashboard_host="0.0.0.0",
+        )
+
+        t.join()
+
+        td = time.time() - time_before_step
+
+        assert 30 < td < 60, (
```
Comment (gjoliver) [inline rllib/tests/test_node_failure.py:212]: self.assertGreater and self.assertLesser man.

**Review by gjoliver — CHANGES_REQUESTED** [review]
very cool, thanks a ton.
I think we need to update the BUILD rule for test_worker_failure.py as well?

File: rllib/tests/test_node_failure.py (hunk 7 above)
Comment (ArturNiederfahrenhorst) [reply rllib/tests/test_node_failure.py:16]: I'm with you, but these constants occur in many core tests and are always not capitalized.
I think consistency > general convention here, right?

File: rllib/tests/test_node_failure.py (hunk 11 above)
Comment (ArturNiederfahrenhorst) [reply rllib/tests/test_node_failure.py:212]: :(

**Comment by sven1977** [comment]
Looks good to me. Just there are a couple of tests failing in the CI. Could you take a quick look and let me know, whether these are this-PR-related?

**Review by sven1977 — APPROVED** [review]
LGTM

**Review by gjoliver — APPROVED** [review]
love these tests!

**Pull request merged**